mm2src/crypto/src/mnemonic.rs

```diff
@@ -0,0 +1,386 @@
+
+    /// The version of the Argon2 algorithm (e.g., 0x13 for the latest version).
+    version: String,
+
+    /// The memory cost parameter defining the memory usage of the algorithm.
+    /// Expressed in kibibytes (KiB).
+    m_cost: u32,
+
+    /// The time cost parameter defining the execution time and number of
+    /// iterations of the algorithm.
+    t_cost: u32,
+
+    /// The parallelism cost parameter defining the number of parallel threads.
+    p_cost: u32,
+}
+
+impl Default for Argon2Params {
+    fn default() -> Self {
+        Argon2Params {
+            algorithm: ARGON2_ALGORITHM.to_string(),
+            version: ARGON2ID_VERSION.to_string(),
+            m_cost: ARGON2ID_M_COST,
+            t_cost: ARGON2ID_T_COST,
+            p_cost: ARGON2ID_P_COST,
+        }
+    }
+}
+
+/// Enum representing different key derivation details.
+///
+/// This enum allows for flexible specification of various key derivation
+/// algorithms and their parameters, making it easier to extend and support
+/// multiple algorithms in the future.
+#[derive(Serialize, Deserialize, Debug)]
+pub enum KeyDerivationDetails {
+    /// Argon2 algorithm with its specific parameters.
+    Argon2(Argon2Params),
+    // Placeholder for future algorithms.
+    // Future algorithms can be added here.
+}
+
+impl Default for KeyDerivationDetails {
+    fn default() -> Self { KeyDerivationDetails::Argon2(Argon2Params::default()) }
+}
+
+/// Represents encrypted mnemonic data for a wallet.
+///
+/// This struct encapsulates all essential components required to securely encrypt
+/// and subsequently decrypt a wallet mnemonic. It is designed to be self-contained,
+/// meaning it includes not only the encrypted data but also all the necessary metadata
+/// and parameters for decryption. This makes the struct portable and convenient for
+/// use in various scenarios, allowing decryption of the mnemonic in different
+/// environments or applications, provided the correct password is supplied.
+///
+/// It includes the following:
+/// - The encryption algorithm used, ensuring compatibility during decryption.
+/// - Detailed key derivation details, including the algorithm and its parameters,
+///   essential for recreating the encryption key from the user's password.
+/// - The Base64-encoded salt for AES key derivation, IV (Initialization Vector),
+///   and the ciphertext itself.
+/// - If HMAC is used, it also includes the salt for HMAC key derivation and the HMAC tag,
+///   which are crucial for ensuring the integrity and authenticity of the encrypted data.
+///
+/// The structure is typically used for wallet encryption in blockchain-based applications,
+/// providing a robust and comprehensive approach to securing sensitive mnemonic data..
+#[derive(Serialize, Deserialize, Debug)]
+pub struct EncryptedMnemonicData {
+    /// The encryption algorithm used to encrypt the mnemonic.
+    /// Example: "AES-256-CBC".
+    encryption_algorithm: String,
```

Comment:
Yeah, enum is better, will fix it. We might add more algorithms in the future for sure if we want to be a framework (SDK). It's not required for GUIs though since we should choose the most secure algorithm for GUIs.

Reply:
Done